File: 19-RemoveNthNodeFromEndOfList/19-RemoveNthNodeFromEndOfList.java
// Last updated: 9/14/2025, 12:55:15 PM
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode fast = head;
        ListNode slow = head;
        int i = 0;
        while(fast.next != null){
            fast = fast.next;
            i++;
            if(i > n) slow = slow.next;
        }
        i++;
        if(n == i) return head.next;
        else slow.next = slow.next.next;
        return head;
    }
}

File: 19-RemoveNthNodeFromEndOfList/19-RemoveNthNodeFromEndOfList.py
# Last updated: 9/14/2025, 12:55:12 PM
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        fast = head
        while n > 0:
            fast = fast.next
            n-=1
        if fast is None:
            return head.next
        slow = head
        while fast.next:
            slow = slow.next
            fast = fast.next
        slow.next = slow.next.next
        return head